system_stats.c

```diff
@@ -47,6 +49,78 @@ static int proc_read_int(const char *name)
 	return ret;
 }
 
+/*
+ * Test if Linux Pressure Stall Information is available. Starting from linux
+ * kernel 4.20 it should manifest itself via /proc/pressure directory in procfs
+ * with cpu,memory,io files.
+ */
+static bool pressure_available()
+{
+	struct stat sb;
+	int res;
+
+	res = stat("/proc/pressure", &sb);
+	return (res != -1) && S_ISDIR(sb.st_mode);
+}
+
+/*
+ * Pressure stall information for specified resource. PSI format for Memory and
+ * IO:
+ *
+ *   some avg10=0.00 avg60=0.00 avg300=0.00 total=0
+ *   full avg10=0.00 avg60=0.00 avg300=0.00 total=0
+ *
+ * For CPU only "some" line is present.
+ *
+ * The result returned via pressure array passed as the first argument.
+ */
+static void read_pressure(pressure *result, pressure_res resource)
+{
+	FILE *f;
+
+	switch (resource)
+	{
+		case CPU:
+			f = fopen(PROC_PRESSURE "cpu", "r");
+			break;
+		case MEMORY:
+			f = fopen(PROC_PRESSURE "memory", "r");
+			break;
+		case IO:
+			f = fopen(PROC_PRESSURE "io", "r");
+			break;
+		default:
+			return;
+	}
+
+	if (f != NULL) {
+		char type[PTYPE_SIZE];
+		pressure p = {0, };
+
+		/* There could be either one or two lines */
+		for (int i = 0; i < 2; i++)
+		{
+			if (fscanf(f, "%s avg10=%f avg60=%f avg300=%f total=%f",
+					   type, &p.avg10, &p.avg60, &p.avg300, &p.total) != 5)
+				break;
+
+			if (strncmp(type, "some", 4) == 0)
+				p.type = SOME;
+
+			if (strncmp(type, "full", 4) == 0)
+				p.type = FULL;
+
```

Comment:
What if we get "full" from the "cpu" (e.g. new kernel)?

Reply:
From what I understand it should not be the case even for new kernels (the [latest](https://www.kernel.org/doc/html/latest/accounting/psi.html?highlight=psi#pressure-interface) says only "the format for CPU ..." with no mentioning it being changed one day). But if needed it could be easily addressed by extending p_cpu array to two elements.